### ISSUE TITLE
Scroll API bugs out many pages in.

### ISSUE THREAD

**Issue by lcpriest** [intercom/intercom-ruby#283]
## Version info
  - intercom-ruby version: 3.5.4
  - Ruby version: 2.3.1

## Expected behavior
The scroll next function grabs the next page of results

## Actual behavior
"The error of type '' is not recognized. It occurred with the message: scroll already exists for this app and http_code: '400'. Please contact Intercom with these details."

## Steps to reproduce
 1. Have more than 40k contacts in an Intercom account
 2. Run client.users.scroll (I do it on initial sync only)
 3. It breaks at random pages

## Notes
Unfortunately, I can't pinpoint the page it occurs at. Though I have a couple of questions.

1. Is the scroll API still rate limited?
2. Is the scroll endpoint an open websocket?
2a) If so, what other ways could scroll be opened twice?
2b) If not, is the gem closing the previous request before opening a new one?

## Logs
```
The error of type '' is not recognized. It occurred with the message: scroll already exists for this app and http_code: '400'. Please contact Intercom with these details.
["/app/vendor/bundle/ruby/2.3.0/gems/intercom-3.5.4/lib/intercom/request.rb:152:in `raise_application_errors_on_failure'", "/app/vendor/bundle/ruby/2.3.0/gems/intercom-3.5.4/lib/intercom/request.rb:94:in `parse_body'", "/app/vendor/bundle/ruby/2.3.0/gems/intercom-3.5.4/lib/intercom/request.rb:70:in `block in execute'", "/app/vendor/ruby-2.3.1/lib/ruby/2.3.0/net/http.rb:853:in `start'", "/app/vendor/bundle/ruby/2.3.0/gems/intercom-3.5.4/lib/intercom/request.rb:65:in `execute'", "/app/vendor/bundle/ruby/2.3.0/gems/intercom-3.5.4/lib/intercom/client.rb:111:in `execute_request'", "/app/vendor/bundle/ruby/2.3.0/gems/intercom-3.5.4/lib/intercom/client.rb:88:in `get'", "/app/vendor/bundle/ruby/2.3.0/gems/intercom-3.5.4/lib/intercom/scroll_collection_proxy.rb:40:in `block in each'", "/app/vendor/bundle/ruby/2.3.0/gems/intercom-3.5.4/lib/intercom/scroll_collection_proxy.rb:38:in `loop'", "/app/vendor/bundle/ruby/2.3.0/gems/intercom-3.5.4/lib/intercom/scroll_collection_proxy.rb:38:in `each'", "/app/lib/channels/intercom_api/service/base_importer_job.rb:11:in `perform'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/execution.rb:34:in `block in perform_now'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:126:in `call'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:506:in `block (2 levels) in compile'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:455:in `call'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:448:in `block (2 levels) in around'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:286:in `block (2 levels) in halting'", "/app/vendor/bundle/ruby/2.3.0/gems/i18n-0.7.0/lib/i18n.rb:257:in `with_locale'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/translation.rb:7:in `block (2 levels) in <module:Translation>'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:391:in `instance_exec'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:391:in `block in make_lambda'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:285:in `block in halting'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:447:in `block in around'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:455:in `call'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:448:in `block (2 levels) in around'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:286:in `block (2 levels) in halting'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/logging.rb:24:in `block (4 levels) in <module:Logging>'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/notifications.rb:164:in `block in instrument'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/notifications/instrumenter.rb:21:in `instrument'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/notifications.rb:164:in `instrument'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/logging.rb:23:in `block (3 levels) in <module:Logging>'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/logging.rb:44:in `block in tag_logger'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/tagged_logging.rb:70:in `block in tagged'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/tagged_logging.rb:26:in `tagged'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/tagged_logging.rb:70:in `tagged'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/logging.rb:44:in `tag_logger'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/logging.rb:20:in `block (2 levels) in <module:Logging>'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:391:in `instance_exec'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:391:in `block in make_lambda'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:285:in `block in halting'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:447:in `block in around'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:455:in `call'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:101:in `__run_callbacks__'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:750:in `_run_perform_callbacks'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:90:in `run_callbacks'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/execution.rb:33:in `perform_now'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/execution.rb:22:in `block in execute'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:126:in `call'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:506:in `block (2 levels) in compile'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:455:in `call'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:448:in `block (2 levels) in around'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:286:in `block (2 levels) in halting'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/railtie.rb:26:in `block (4 levels) in <class:Railtie>'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/execution_wrapper.rb:76:in `wrap'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/reloader.rb:68:in `block in wrap'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/execution_wrapper.rb:76:in `wrap'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/reloader.rb:67:in `wrap'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/railtie.rb:25:in `block (3 levels) in <class:Railtie>'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:391:in `instance_exec'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:391:in `block in make_lambda'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:285:in `block in halting'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:447:in `block in around'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:455:in `call'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:101:in `__run_callbacks__'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:750:in `_run_execute_callbacks'", "/app/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/callbacks.rb:90:in `run_callbacks'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/execution.rb:20:in `execute'", "/app/vendor/bundle/ruby/2.3.0/gems/activejob-5.0.0.1/lib/active_job/queue_adapters/sidekiq_adapter.rb:40:in `perform'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:152:in `execute_job'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:134:in `block (2 levels) in process'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/chain.rb:128:in `block in invoke'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/server/active_record.rb:6:in `call'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/chain.rb:130:in `block in invoke'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/server/retry_jobs.rb:74:in `call'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/chain.rb:130:in `block in invoke'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/server/logging.rb:11:in `block in call'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/logging.rb:32:in `with_context'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/server/logging.rb:7:in `call'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/chain.rb:130:in `block in invoke'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/middleware/chain.rb:133:in `invoke'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:129:in `block in process'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:168:in `stats'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:128:in `process'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:80:in `process_one'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/processor.rb:68:in `run'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/util.rb:17:in `watchdog'", "/app/vendor/bundle/ruby/2.3.0/gems/sidekiq-4.1.2/lib/sidekiq/util.rb:25:in `block in safe_thread'"]
```

**Comment by choran** [comment]
Hey @lcpriest,
1/ Yep, the scroll API is still rate limited.
2/ No, it is not a websocket. 
2b/ Are you asking if it is making a new request each time? then yes, for every 100 users the scroll will make a new request
e.g. intercom.users.scroll.each { |user| puts user.name}
if you have 1000 users then this will make 10 requests. At the rate it makes requests I would not have thought it would hit any rate limit. But if you were running other requests in parallel then I think there may be the potential for issues then alright.

The error you are seeing indicates that the request tried to access the scroll requesting a new scroll_param. When you make the each request it asks for a new scroll_param and then uses that on subsequent calls to iterate through the list. If you stop that process and then restart it you will see the same error.

I was trying to reproduce it here on a a large data set but could not. Was there anything that could have caused it to make a new request? i.e. a new each call?

Can you provide a code snippet of the code you are using when this occurs? I know you said it occurs on different pages but was wondering if it occurs every time?

**Comment by lcpriest** [comment]
Thank you for all of that information!
  

  
No, there are no other requests in parallel. The rate limit was a one time thing so I think it's safe to ignore.
  

  
The scroll parameters error happens every time. I have an each block on users.scroll that saves a contact if it doesn't exist and moves on if it does, it tends to get am extra 2000 users each time before the scroll param gets lost and it throws that error. It currently occurs every time as I still have not yet fetched all of the users.
  

  
The code is essentially   
  
```
  
client.users.scroll.each do |user|
  
unless array.include?(user)
  
User.create(parsed user)
  
end
  
```
  
  

  
  
>

**Comment by flood4life** [comment]
@lcpriest is it possible that your block could spend more than 1 minute to process a 100 users? E.g., `array.include?` is O(n) and your array is big, or complex logic around `User.create`.

**Comment by lcpriest** [comment]
That was a contrived example, it's not actually an array, but I don't think so. What error message would return if the scroll parameter timed out?
  
  

  
  
>

**Comment by choran** [comment]
I think it should be ...
Intercom::ResourceNotFound: scroll parameter not found

**Comment by choran** [comment]
But @flood4life has a point, have you tried to list all of the users without doing any operation on them? just to see if it completes?

**Comment by lcpriest** [comment]
Ah. That might be it then. I'll try it again and just append to a file instead and see if it breaks.
  
  
  

  
  
>

**Comment by lcpriest** [comment]
I'll try that! There are 85k users, will I have to rate limit or will it delay itself?
  
  
  

  
  
>

**Comment by flood4life** [comment]
@lcpriest it won't hit limit, don't worry. We process a >800k user list every day, and it works fine.

**Comment by lcpriest** [comment]
Alright! You can close or put this issue on hold until I have more information for you :)
  
  

  
  
>

**Comment by choran** [comment]
👍 
cool. Please re-open it when you have more info or any other question

**Comment by lcpriest** [comment]
For closure, I looped through 85k and just printed an incrementing number. Worked fine.

**Comment by choran** [comment]
great. thanks for the update

**Comment by glebe** [comment]
Hi guys, just FYI i'm having the same issue when using intercom.users.scroll.each

Can't find a pattern on how it gets an error, but if you want me to run any tests, i'd be happy to help.

>scroll parameter not found - ["/project/shared/bundle/ruby/2.2.0/gems/intercom-3.5.23/lib/intercom/request.rb:159
